# Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd

In [2]:
df_horses = pd.read_csv('horse_data.csv', names = ['surgery', 'age','HospitalNumber', 'RectalTemperature', 'pulse', 'RespiratoryRate','TemperatureOfExtremities',
             'peripheralPulse',' mucousMembranes', 'capillaryRefillTime', 'pain', 'peristalsis', 'abdominalDistension',
             'nasogastricTube','nasogastricReflux', 'nasogastricRefluxPH','rectalExamination','abdomen','packedCellVolume',
              'totalProtein','abdominocentesisAppearance','abdomcentesisTotalProtein','outcome', 'surgicalLesion', 'typeOfLesion',
              'typeOfLesion_1', 'typeOfLesion_2','cpData'], na_values='?')
df_horses.head()

#параметр na_values='?' заменяет ? на Nan

surgery  age  HospitalNumber  RectalTemperature  pulse  RespiratoryRate  \
0      2.0    1          530101               38.5   66.0             28.0   
1      1.0    1          534817               39.2   88.0             20.0   
2      2.0    1          530334               38.3   40.0             24.0   
3      1.0    9         5290409               39.1  164.0             84.0   
4      2.0    1          530255               37.3  104.0             35.0   

   TemperatureOfExtremities  peripheralPulse   mucousMembranes  \
0                       3.0              3.0               NaN   
1                       NaN              NaN               4.0   
2                       1.0              1.0               3.0   
3                       4.0              1.0               6.0   
4                       NaN              NaN               6.0   

   capillaryRefillTime  ...  packedCellVolume  totalProtein  \
0                  2.0  ...              45.0           8.4   
1                  1.0  ...              50.0          85.0   
2                  1.0  ...              33.0           6.7   
3                  2.0  ...              48.0           7.2   
4                  2.0  ...              74.0           7.4   

   abdominocentesisAppearance  abdomcentesisTotalProtein  outcome  \
0                         NaN                        NaN      2.0   
1                         2.0                        2.0      3.0   
2                         NaN                        NaN      1.0   
3                         3.0                        5.3      2.0   
4                         NaN                        NaN      2.0   

   surgicalLesion  typeOfLesion  typeOfLesion_1  typeOfLesion_2  cpData  
0               2         11300               0               0       2  
1               2          2208               0               0       2  
2               2             0               0               0       1  
3               1          2208               0               0       1  
4               2          4300               0               0       2  

[5 rows x 28 columns]

In [3]:
df_horses_part = df_horses[['surgery','age', 'RectalTemperature','pulse', 'RespiratoryRate','abdominalDistension', 'totalProtein', 'outcome']]
df_horses_part.head()

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
0      2.0    1               38.5   66.0             28.0   
1      1.0    1               39.2   88.0             20.0   
2      2.0    1               38.3   40.0             24.0   
3      1.0    9               39.1  164.0             84.0   
4      2.0    1               37.3  104.0             35.0   

   abdominalDistension  totalProtein  outcome  
0                  4.0           8.4      2.0  
1                  2.0          85.0      3.0  
2                  1.0           6.7      1.0  
3                  4.0           7.2      2.0  
4                  NaN           7.4      2.0

Определим, как распределяются значения столбца Операция:

In [4]:
surgery_counts = {}

for surgery in df_horses_part.surgery.round():
    if surgery not in surgery_counts:
        surgery_counts[surgery] = 1
    else:
        surgery_counts[surgery] += 1
        
surgery_counts    

{2.0: 119, 1.0: 180, nan: 1}

Данные по столбцу Операция интерпретируются следующим образом: 1 = Да, была операция, 2 = Лечение прошло без хирургического вмешательства. Наблюдаем пропуск в данных.        

Определим, как распределяются значения столбца Возраст.

In [5]:
age_counts = {}

for age in df_horses_part.age.round():
    if age not in age_counts:
        age_counts[age] = 1
    else:
        age_counts[age] += 1
        
age_counts        
        

{1: 276, 9: 24}

Согласно описания Базы данных по конским коликам: 1 = взрослая лошадь, 2 = молодая (< 6 месяцев). По полученным результатам 1 встречается 276 раз, 9 - 24 раза. Будем считать, что 9 это опечатка в данных (вместо цифры 9 должны были проставит 2).


Определим, как распределяются значения столбца Вздутие живота:

In [6]:
abdominalDistension_counts = {}

for abdominalDistension in df_horses_part.abdominalDistension.round():
    if abdominalDistension not in abdominalDistension_counts:
        abdominalDistension_counts[abdominalDistension] = 1
    else:
        abdominalDistension_counts[abdominalDistension] += 1
        
abdominalDistension_counts    

{4.0: 38,
 2.0: 65,
 1.0: 76,
 nan: 1,
 3.0: 65,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1,
 nan: 1}

Данные столбца Вздутие живота интерпретируются следующим образом: 1 = нет, 2 = легкий, 3 = умеренный, 4 = тяжелый. В данных по этому столбцу наблюдаем много пропусков (из 300 заполнено 244).                

In [7]:
df_horses_part.describe()

surgery         age  RectalTemperature       pulse  RespiratoryRate  \
count  299.000000  300.000000         240.000000  276.000000       242.000000   
mean     1.397993    1.640000          38.167917   71.913043        30.417355   
std      0.490305    2.173972           0.732289   28.630557        17.642231   
min      1.000000    1.000000          35.400000   30.000000         8.000000   
25%      1.000000    1.000000          37.800000   48.000000        18.500000   
50%      1.000000    1.000000          38.200000   64.000000        24.500000   
75%      2.000000    1.000000          38.500000   88.000000        36.000000   
max      2.000000    9.000000          40.800000  184.000000        96.000000   

       abdominalDistension  totalProtein     outcome  
count           244.000000    267.000000  299.000000  
mean              2.266393     24.456929    1.551839  
std               1.065131     27.475009    0.737187  
min               1.000000      3.300000    1.000000  
25%               1.000000      6.500000    1.000000  
50%               2.000000      7.500000    1.000000  
75%               3.000000     57.000000    2.000000  
max               4.000000     89.000000    3.000000

In [8]:
print('Значение моды для столбца Температура',df_horses_part.RectalTemperature.round().mode()[0])

print('Значение моды для столбца Пульс',df_horses_part.pulse.round().mode()[0] )

print('Значение моды для столбца Частота дыхания',df_horses_part.RespiratoryRate.round().mode()[0] )

print('Значение моды для столбца Общий белок',df_horses_part.totalProtein.round().mode()[0] )

print('Значение моды для столбца Возраст',df_horses_part.age.round().mode()[0] )

Значение моды для столбца Температура 38.0
Значение моды для столбца Пульс 48.0
Значение моды для столбца Частота дыхания 20.0
Значение моды для столбца Общий белок 7.0
Значение моды для столбца Возраст 1


Определим, как распределяются значения столбца Исход.

In [9]:
outcome_counts = {}

for outcome in df_horses_part.outcome.round():
    if outcome not in outcome_counts:
        outcome_counts[outcome] = 1
    else:
        outcome_counts[outcome] += 1
        
outcome_counts       

{2.0: 77, 3.0: 44, 1.0: 178, nan: 1}

Значения столбца Исход описывают, что в итоге случилось с лошадью. 1 = жил, 2 = умер, 3 = был усыплен.
               

   Датафрейм df_horses_part представляет собой набор данных, характеризующих физические показатели состояния здоровья 300 лошадей, страдающих кишечными коликами.Среди них есть как качественные (возрастная категория, хирургическая операция была/нет), так и количественные (пульс, частота дыхания, показатель белка). В имеющихся данных присутствует большое количество пропусков по многим показателям, особеннов данных о температуре, частоте дыхания, категории вздутия живота. Согласно данным, в большинстве случаев конскими коликами страдают чаще взрослые особи 276 из 300, что составляет 92%. Больше половины (60%) в ходе лечения подверглась хирургическому вмешательству.   
   Воспольземся методом .describe() для проведения описательной статистики количественных показателей. Так, при анализе выборки в 300 лошадей было определено, что средняя температура составляет 38.167917 градусов, медиана 38,2, мода 38,00 (нормальная температура 37,8).
   Среднее значение пульса 71.913043, при СКО 28.630557, что связано с большим размахом данных от частоты в 30 до 184 ударов. Мода и медиана по столбцу Пульс составила 48 и 64 соответственно. Нормальная частота ударов для взрослой лошади 30-40 ударов в минуту.
   Среднее значение Частоты дыхания составило 30.417355, также с довольно большим показателем СКО 17.642231 при размахе данных от 8 до 96. Мода 20.0, медиана 24.50. Нормальная оценка от 8 до 10.
   Среднее значение показателя Общий белок составило 24.456929, СКО 27.475009. Минимальное значение 3.300000, максимальное значение 89.000000. Мода 7.0, медиана 7.500000. Нормальные значения лежат в диапазоне 6-7,5 (г / дл)
   Кроме того, еще одним анализируемым показателем является вздутие живота. По имеющимся данным наиболее часто его нет совсем, часто легкое или умеренное.
   Исход болезни: 178 лошадей из 300 (60%) выздоровели, остальные умерли или был усыплены.


# Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

Найдем выбросы в столбце RectalTemperature:

In [10]:
print(df_horses_part.RectalTemperature.mean())
print(df_horses_part[(df_horses_part.RectalTemperature > 35) & (df_horses_part.RectalTemperature < 42)].RectalTemperature.mean())

38.16791666666669
38.16791666666669


По столбцу RectalTemperature выбросов не наблюдается, данные в пределах реально допустимых показателей температуры тела.

Найдем выбросы по столбцу pulse через межквартильный размах:

In [18]:
q1 = df_horses_part.pulse.quantile(0.25)
q3 = df_horses_part.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df_horses_part[df_horses_part.pulse.between(lower_bound, upper_bound, inclusive=True)]  #inclusive=True это включительно
remove_outliers

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
0        2.0    1               38.5   66.0             28.0   
1        1.0    1               39.2   88.0             20.0   
2        2.0    1               38.3   40.0             24.0   
4        2.0    1               37.3  104.0             35.0   
6        1.0    1               37.9   48.0             16.0   
..       ...  ...                ...    ...              ...   
295      1.0    1                NaN  120.0             70.0   
296      2.0    1               37.2   72.0             24.0   
297      1.0    1               37.5   72.0             30.0   
298      1.0    1               36.5  100.0             24.0   
299      1.0    1               37.2   40.0             20.0   

     abdominalDistension  totalProtein  outcome  
0                    4.0           8.4      2.0  
1                    2.0          85.0      3.0  
2                    1.0           6.7      1.0  
4                    NaN           7.4      2.0  
6                    3.0           7.0      1.0  
..                   ...           ...      ...  
295                  NaN          65.0      3.0  
296                  3.0           NaN      3.0  
297                  3.0           6.8      2.0  
298                  3.0           6.0      1.0  
299                  NaN          62.0      3.0  

[271 rows x 8 columns]

Посмотрим строки, которые содержат данные, которые мы принимаем за выбросы


In [15]:
pd.concat([df_horses_part, remove_outliers]).drop_duplicates(keep=False)

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
3        1.0    9               39.1  164.0             84.0   
5        2.0    1                NaN    NaN              NaN   
28       1.0    1                NaN    NaN              NaN   
41       2.0    9               39.0  150.0             72.0   
52       2.0    1                NaN    NaN              NaN   
55       1.0    9               38.6  160.0             20.0   
56       1.0    1                NaN    NaN              NaN   
58       1.0    1                NaN    NaN             20.0   
74       1.0    9                NaN    NaN              NaN   
93       2.0    1                NaN    NaN              NaN   
115      2.0    1                NaN    NaN             40.0   
117      2.0    1               39.5    NaN              NaN   
126      1.0    1               38.7    NaN              NaN   
151      1.0    1               39.3    NaN              NaN   
159      2.0    1                NaN    NaN             12.0   
160      1.0    1                NaN    NaN              NaN   
174      2.0    1                NaN    NaN              NaN   
204      1.0    1               39.2    NaN             23.0   
207      1.0    1                NaN    NaN              NaN   
216      2.0    1                NaN    NaN              NaN   
227      1.0    9                NaN    NaN              NaN   
255      1.0    9               38.8  184.0             84.0   
275      1.0    9               38.8  150.0             50.0   

     abdominalDistension  totalProtein  outcome  
3                    4.0           7.2      2.0  
5                    2.0           NaN      1.0  
28                   NaN           NaN      2.0  
41                   NaN           8.5      1.0  
52                   1.0           7.7      1.0  
55                   4.0           NaN      2.0  
56                   NaN           6.7      1.0  
58                   3.0           5.9      2.0  
74                   NaN           4.9      2.0  
93                   3.0           6.5      2.0  
115                  2.0          70.0      1.0  
117                  3.0           6.7      3.0  
126                  2.0           6.5      1.0  
151                  2.0           NaN      2.0  
159                  2.0           7.5      1.0  
160                  NaN          58.0      1.0  
174                  4.0           NaN      2.0  
204                  2.0           6.6      1.0  
207                  NaN           8.7      2.0  
216                  NaN           NaN      1.0  
227                  NaN           6.5      1.0  
255                  3.0           3.3      2.0  
275                  2.0           6.2      2.0

Таким образом, нашли строки, в которых пульс от 150 до 184 ударов при норме 30-40. Эти данные оставляем, так как мы исследуем данные о состоянии больных животных, которые могут имет повышенную частоту сердечных сокращений.


Найдем выбросы по столбцу RespiratoryRate

In [28]:
q1 = df_horses_part.RespiratoryRate.quantile(0.25)
q3 = df_horses_part.RespiratoryRate.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers_ = df_horses_part[df_horses_part.RespiratoryRate.between(lower_bound, upper_bound, inclusive=True)]  #inclusive=True это включительно
remove_outliers_

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
0        2.0    1               38.5   66.0             28.0   
1        1.0    1               39.2   88.0             20.0   
2        2.0    1               38.3   40.0             24.0   
4        2.0    1               37.3  104.0             35.0   
6        1.0    1               37.9   48.0             16.0   
..       ...  ...                ...    ...              ...   
294      2.0    1               38.5   40.0             16.0   
296      2.0    1               37.2   72.0             24.0   
297      1.0    1               37.5   72.0             30.0   
298      1.0    1               36.5  100.0             24.0   
299      1.0    1               37.2   40.0             20.0   

     abdominalDistension  totalProtein  outcome  
0                    4.0           8.4      2.0  
1                    2.0          85.0      3.0  
2                    1.0           6.7      1.0  
4                    NaN           7.4      2.0  
6                    3.0           7.0      1.0  
..                   ...           ...      ...  
294                  1.0          67.0      1.0  
296                  3.0           NaN      3.0  
297                  3.0           6.8      2.0  
298                  3.0           6.0      1.0  
299                  NaN          62.0      3.0  

[225 rows x 8 columns]

In [29]:
pd.concat([df_horses_part, remove_outliers_]).drop_duplicates(keep=False).head()

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
3       1.0    9               39.1  164.0             84.0   
5       2.0    1                NaN    NaN              NaN   
7       1.0    1                NaN   60.0              NaN   
9       2.0    9               38.3   90.0              NaN   
28      1.0    1                NaN    NaN              NaN   

    abdominalDistension  totalProtein  outcome  
3                   4.0           7.2      2.0  
5                   2.0           NaN      1.0  
7                   2.0           8.3      2.0  
9                   1.0           6.2      1.0  
28                  NaN           NaN      2.0

Нормальная частота дыхания взрослой лошади 7-10 раз. Больныеживотные могут иметь учащенное дыхание, поэтому все, что определили как выбросы, оставляем в данных.


Найдем выбросы по столбцу totalProtein.

In [26]:
q1 = df_horses_part.totalProtein.quantile(0.25)
q3 = df_horses_part.totalProtein.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers_1 = df_horses_part[df_horses_part.totalProtein.between(lower_bound, upper_bound, inclusive=True)]  #inclusive=True это включительно
remove_outliers_1

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
0        2.0    1               38.5   66.0             28.0   
1        1.0    1               39.2   88.0             20.0   
2        2.0    1               38.3   40.0             24.0   
3        1.0    9               39.1  164.0             84.0   
4        2.0    1               37.3  104.0             35.0   
..       ...  ...                ...    ...              ...   
294      2.0    1               38.5   40.0             16.0   
295      1.0    1                NaN  120.0             70.0   
297      1.0    1               37.5   72.0             30.0   
298      1.0    1               36.5  100.0             24.0   
299      1.0    1               37.2   40.0             20.0   

     abdominalDistension  totalProtein  outcome  
0                    4.0           8.4      2.0  
1                    2.0          85.0      3.0  
2                    1.0           6.7      1.0  
3                    4.0           7.2      2.0  
4                    NaN           7.4      2.0  
..                   ...           ...      ...  
294                  1.0          67.0      1.0  
295                  NaN          65.0      3.0  
297                  3.0           6.8      2.0  
298                  3.0           6.0      1.0  
299                  NaN          62.0      3.0  

[267 rows x 8 columns]

In [27]:
pd.concat([df_horses_part, remove_outliers_1]).drop_duplicates(keep=False).head()

surgery  age  RectalTemperature  pulse  RespiratoryRate  \
5       2.0    1                NaN    NaN              NaN   
17      2.0    1               37.5   48.0             24.0   
25      2.0    1               37.8   60.0             42.0   
28      1.0    1                NaN    NaN              NaN   
39      1.0    9               39.2  146.0             96.0   

    abdominalDistension  totalProtein  outcome  
5                   2.0           NaN      1.0  
17                  NaN           NaN      1.0  
25                  NaN           NaN      1.0  
28                  NaN           NaN      2.0  
39                  NaN           NaN      2.0

Нормальный показатель белка 6.5-7.0. У больных животных он может быть повышен, поэтому оставляем в данных.

# Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [ ]:
df_horses_part.info()

In [ ]:
for col in df_horses_part.columns:                       #считаем пропуски в данных
    pct_missing = df_horses_part[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

В столбце surgery наблюдается только один пропуск об отсутствии/наличии операции. Его можем заменить на константу. 

In [ ]:
df_horses_part_const_surg = df_horses_part['surgery'].fillna('no_info')

In [ ]:
df_horses_part_const_surg.unique()             
    

Аналогично можно заменить пропуск в столбце outcome (исход)

In [ ]:
df_horses_part_const_out = df_horses_part['outcome'].fillna('no_info')
df_horses_part_const_out.unique() 

Столбец abdominalDistension(вздутие живота) содержит качественные, порядковые данные. С ними не провести никаких агрегированных расчетов, поэтому пропуски также можно выделить в отдельную категорию no_info

In [ ]:
df_horses_part_const_aD = df_horses_part['abdominalDistension'].fillna('no_info')
df_horses_part_const_aD.unique() 

В столбце RectalTemperature не заполнено 20% данных о температуре тела животных. Среднеарифметическое значение - 38.167917, медиана - 38.200000, мода - 38.0. Значения трех показателей довольно близки друг к другу, поэтому если заменим пропуски среднеарифметическим значением, это не приведет к сильному искажению данных.

In [ ]:
print(df_horses_part.RectalTemperature.mean())
print(df_horses_part.RectalTemperature.median())
print(df_horses_part.RectalTemperature.mode()[0])
print(df_horses_part.RectalTemperature.std())

In [ ]:
df_mean = pd.read_csv('horse_data.csv', names = ['surgery', 'age','HospitalNumber', 'RectalTemperature', 'pulse', 'RespiratoryRate',
                   'TemperatureOfExtremities','peripheralPulse',' mucousMembranes', 'capillaryRefillTime', 'pain', 'peristalsis', 'abdominalDistension',
             'nasogastricTube','nasogastricReflux', 'nasogastricRefluxPH','rectalExamination','abdomen','packedCellVolume',
              'totalProtein','abdominocentesisAppearance','abdomcentesisTotalProtein','outcome', 'surgicalLesion', 'typeOfLesion',
              'typeOfLesion_1', 'typeOfLesion_2','cpData'], na_values='?')
df_mean.RectalTemperature.fillna(df_horses_part.RectalTemperature.mean(), inplace=True)
df_mean.RectalTemperature.fillna(df_horses_part.RectalTemperature.mean()).isna().sum()

In [ ]:
print(df_mean.RectalTemperature.mean())
print(df_mean.RectalTemperature.median())
print(df_mean.RectalTemperature.mode()[0])
print(df_mean.RectalTemperature.std())


В столбце pulse показатели моды, мендианы, среднего значения довольно сильно отличаются друг от друга, поэтому предполагаю, что лучше заменить пропуски показателем моды, как наиболее часто встречающимся значением.

In [ ]:
print(df_horses_part.pulse.mean())
print(df_horses_part.pulse.median())
print(df_horses_part.pulse.mode()[0])
print(df_horses_part.pulse.std())

In [ ]:
df_moda_pulse = pd.read_csv('horse_data.csv', names = ['surgery', 'age','HospitalNumber', 'RectalTemperature', 'pulse', 'RespiratoryRate',
                   'TemperatureOfExtremities','peripheralPulse',' mucousMembranes', 'capillaryRefillTime', 'pain', 'peristalsis', 'abdominalDistension',
             'nasogastricTube','nasogastricReflux', 'nasogastricRefluxPH','rectalExamination','abdomen','packedCellVolume',
              'totalProtein','abdominocentesisAppearance','abdomcentesisTotalProtein','outcome', 'surgicalLesion', 'typeOfLesion',
              'typeOfLesion_1', 'typeOfLesion_2','cpData'], na_values='?')

In [ ]:
df_moda_pulse.pulse.fillna(df_horses_part.pulse.mode()[0], inplace=True)
df_moda_pulse.pulse.fillna(df_horses_part.pulse.mode()[0]).isna().sum()

In [ ]:
print(df_moda_pulse.pulse.mean())
print(df_moda_pulse.pulse.median())
print(df_moda_pulse.pulse.mode()[0])
print(df_moda_pulse.pulse.std())


Аналогично для столбцов RespiratoryRate и totalProtein

In [ ]:
print(df_horses_part.RespiratoryRate.mean())
print(df_horses_part.RespiratoryRate.median())
print(df_horses_part.RespiratoryRate.mode()[0])
print(df_horses_part.RespiratoryRate.std())

In [ ]:
df_moda_RespR = pd.read_csv('horse_data.csv', names = ['surgery', 'age','HospitalNumber', 'RectalTemperature', 'pulse', 'RespiratoryRate',
                   'TemperatureOfExtremities','peripheralPulse',' mucousMembranes', 'capillaryRefillTime', 'pain', 'peristalsis', 'abdominalDistension',
             'nasogastricTube','nasogastricReflux', 'nasogastricRefluxPH','rectalExamination','abdomen','packedCellVolume',
              'totalProtein','abdominocentesisAppearance','abdomcentesisTotalProtein','outcome', 'surgicalLesion', 'typeOfLesion',
              'typeOfLesion_1', 'typeOfLesion_2','cpData'], na_values='?')

df_moda_RespR.RespiratoryRate.fillna(df_horses_part.RespiratoryRate.mode()[0], inplace=True)
df_moda_RespR.RespiratoryRate.fillna(df_horses_part.RespiratoryRate.mode()[0]).isna().sum()


In [ ]:
print(df_moda_RespR.RespiratoryRate.mean())
print(df_moda_RespR.RespiratoryRate.median())
print(df_moda_RespR.RespiratoryRate.mode()[0])
print(df_moda_RespR.RespiratoryRate.std())


In [ ]:
print(df_horses_part.totalProtein.mean())
print(df_horses_part.totalProtein.median())
print(df_horses_part.totalProtein.mode()[0])
print(df_horses_part.totalProtein.std())

In [ ]:
df_moda_totalProtein = pd.read_csv('horse_data.csv', names = ['surgery', 'age','HospitalNumber', 'RectalTemperature', 'pulse', 'RespiratoryRate',
                   'TemperatureOfExtremities','peripheralPulse',' mucousMembranes', 'capillaryRefillTime', 'pain', 'peristalsis', 'abdominalDistension',
             'nasogastricTube','nasogastricReflux', 'nasogastricRefluxPH','rectalExamination','abdomen','packedCellVolume',
              'totalProtein','abdominocentesisAppearance','abdomcentesisTotalProtein','outcome', 'surgicalLesion', 'typeOfLesion',
              'typeOfLesion_1', 'typeOfLesion_2','cpData'], na_values='?')

df_moda_totalProtein.totalProtein.fillna(df_horses_part.totalProtein.mode()[0], inplace=True)
df_moda_totalProtein.totalProtein.fillna(df_horses_part.totalProtein.mode()[0]).isna().sum()

In [ ]:
print(df_moda_totalProtein.totalProtein.mean())
print(df_moda_totalProtein.totalProtein.median())
print(df_moda_totalProtein.totalProtein.mode()[0])
print(df_moda_totalProtein.totalProtein.std())